README.md
# 🎨 Flutter Spinwheel

A widget that outputs items (text or image) along a pan-able circular wheel/spinner with customization options.

## 🍀 Features

- Auto-play (enable/disable)
- Long-press to pause (enable/disable)
- Size adjustments
- Text/Image support
- Image tweaking support
- Clockwise and counter-clockwise pan to navigate
- Touch to navigate in previously panned direction
- Paint customization to alter the look
- Callback function to notify selected item

## 📥 Install & Import

Add:

```yaml
dependencies:
  flutter_spinwheel: "^1.0.0"
```

among your other dependencies in pubspec.yaml and then import it in your required project:

```dart
import  'package:flutter_spinwheel/flutter_spinwheel.dart';
```

## 🔧 Usage

```dart
List<String> items = ['🏘',  '🚓',  '🚛',  '🏍',  '🎉'];

List<NamedImage> imgPack = [
	NamedImage(
		path:  'assets/images/car.jpg',
		name:  'car',
	),
	NamedImage(
		path:  'assets/images/bike.jpg',
		name:  'bike',
	),
	NamedImage(
		path:  'assets/images/other.jpg',
		name:  'other',
	),
];
```

```dart
final spinwheelWithText = Spinwheel(
	items: items,
	onChanged: (val) {
	if (this.mounted)
		setState(() {
			currentText = val;
		});
	},
	size:  150.0,
	select:  2,
	autoPlay:  true,
);
```

```dart
final spinwheelWithImage = Spinwheel(
	size:  250.0,
	imageItems: imgPack,
	select:  1,
	onChanged: (val) {
		carousel.animateToPage(carouselNames.indexOf(val),
		duration:  Duration(seconds:  1), curve:  Curves.linear);
	},
	rotationDuration:  250,
	autoPlay:  true,
	longPressToPauseAutoplay:  true,
	hideOthers:  false,
	shouldDrawBorder:  false,
	shouldDrawCenterPiece:  false,
	wheelOrientation: pi /  4,
);
```

## 📁 Sample

<table>
    <tr>
        <td align="center">
            <img  src="https://raw.githubusercontent.com/jerilMJ/flutter_spinwheel/master/screenshots/emoji_demo.gif"  width="500px">
            Text Demo
        </td>
        <td align="center">
            <img  src="https://raw.githubusercontent.com/jerilMJ/flutter_spinwheel/master/screenshots/image_demo.gif"  width="500px">
            Image Demo
        </td>
        <td align="center">
            <img  src="https://raw.githubusercontent.com/jerilMJ/flutter_spinwheel/master/screenshots/sample.gif"  width="500px">
            Simple Demo
        </td>
    </tr>
</table>

## 🚀 Future Plans

- Tap on any item on the spinner to rotate to that item
- More fluid animations

## 👨🏼‍💻 Support / Feature Request / Bug Report

If you experience any issues or need help understanding some of the functionalities or you feel that it can be improved, feel free to open an issue on the repo (ofcourse, first look if there is already a similar open issue).


CHANGELOG.md
## [0.1.0] - 02/04/2020

- Fixed some sizing issues

## [0.0.2] - 31/03/2020

- Release to pub.dev and testing

## [0.0.1] - 30/03/2020

- Initial release for testing


test/flutter_spinwheel_test.dart
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('no test', () {});
}


lib/named_image.dart
import 'package:flutter/material.dart';

/// Custom class used with Spinwheel to distinguish between images
/// and customize individual settings of images.
class NamedImage {
  /// Take note that [path] should be set to the full path name just like what
  /// you would write in pubspec.yaml.
  ///
  /// IMPORTANT!! The [name] field is the field that will be returned in the callback
  /// in the [onChanged] field of the Spinheel class.
  ///
  /// Change [offsetX] (default: 1) to horizontally offset.
  /// Change [offsetY] (default: 1.5) to vertically offset.
  /// [fit] describes the bounding box of the image.
  NamedImage({
    @required this.path,
    @required this.name,
    this.offsetX = 1,
    this.offsetY = 1.5,
    this.fit = BoxFit.cover,
    this.filter,
  });

  /// Path to the image file (same as what you would describe in pubspec.yaml).
  String path;

  /// Name of the image to be used as callback value.
  String name;

  /// Horizontally offset image.
  double offsetX;

  /// Vertically offset image.
  double offsetY;

  /// Bounding box of the image.
  BoxFit fit;

  /// Filter to use on image. By default, none. Refer to ColorFilter doc
  /// information on how to use.
  ColorFilter filter;
}


lib/src/spinwheel.dart
import 'dart:ui' as ui;
import 'dart:async';

import 'package:flutter/material.dart';

import 'spinwheel_painter.dart';
import 'named_image.dart';
import 'utilities.dart';

typedef VoidPassValue = void Function(String value);

/// Custom-made UI widget that takes an array of strings as [items] and makes
/// a wheel with each sector of the wheel assigned to each of the option. Use the [select]
/// param to choose which sector of the wheel to act as the selecting sector. Use the [size]
/// param to specify the area to be occupied by the widget.
///
/// This widget uses the CustomPainter class to draw itself and its component on the given
/// area specified by [size]. Some customizations are provided as params. For additional
/// customizations, you will have to edit the source code and tweak the SpinwheelPainter
/// to your liking.
class Spinwheel extends StatefulWidget {
  /// The [size] of the square on which the circular spinner will be drawn.
  final double size;

  /// The [items] that will be shown on the spinner. Each will get an equally
  /// separated sector of the circle. (360° / no. of items) .Either this or [imageItems] must not
  /// be null for this widget to work.
  final List<String> items;

  /// The [imageItems] to display on the spinner. Either this or [items] must not
  /// be null for this widget to work.
  final List<NamedImage> imageItems;

  /// The position of the selection (hihglighted) sector of the circle.
  /// The range is : 0..(size of items).. Imagine it just like an array.
  /// Selection numbering starts from 0.
  final int select;

  /// Callback that returns the selected value from spinner menu everytime the
  /// selection is changed.
  final VoidPassValue onChanged;

  /// The number of menu-items given to the spinwheel class.
  /// Should be handled in the constructor.
  final int itemCount;

  /// The duration (in milliseconds) of the rotation animation.
  final int rotationDuration;

  /// Decides whether touching the spinner will make it rotate in the
  /// previously panned direction (default-clockwise).
  final bool touchToRotate;

  /// Set to true to autoplay.
  final bool autoPlay;

  /// Default duration is 5 seconds.
  final Duration autoPlayDuration;

  /// Set to true to cancel autoplay temporarily when touching.
  final bool longPressToPauseAutoplay;

  /// Paint settings used for drawing wheel. Customizable
  final Paint wheelPaint;

  /// Paint settings used for drawing wheel border if applicable. Customizable
  final Paint borderPaint;

  /// Paint settings used for drawing the dividers if applicable. Customizable
  final Paint sectorDividerPaint;

  /// Paint settings used for drawing the center piece (circle) if applicable. Customizable
  final Paint centerPiecePaint;

  /// Paint settings used for drawing the highlight if applicable. Customizable
  final Paint highlightPaint;

  /// Paint settings used for drawing the shutter if applicable. Customizable
  final Paint shutterPaint;

  /// Determines whether sector dividers should be drawn.
  final bool shouldDrawDividers;

  /// Determines whether border should be drawn.
  final bool shouldDrawBorder;

  /// Determines whether center piece should be drawn.
  final bool shouldDrawCenterPiece;

  /// Determines whether highlight should be drawn.
  final bool shouldHighlight;

  /// Determines whether the shutter to hide all options except selected should be drawn.
  final bool hideOthers;

  /// Determines whether highlighter should be present while rotating.
  /// This might seem not so good on some designs.
  final bool highlightWhileRotating;

  /// The angle to rotate the wheel by (around center).
  final double wheelOrientation;

  /// The duration for fade-in of the spinner. Set to Duration(unit: 0) for disabling.
  final Duration fadeDuration;

  /// This widget is displayed while the images for the spinner are being loaded.
  /// Default load indicator is a text widget with three dots.
  final Widget loadingIndicator;

  /// Creates a SpinnerWheel (sort of like a radial menu) which takes an items list (list of strings [items])
  /// or images list (list of images [imageItems]). It provides callback [onChanged] to keep track of
  /// the current value and many other customization options (which are given as parameters of type [Paint]).
  /// If the provided customizations feel short of your needs, please edit the source code according to
  /// your demands.
  ///
  /// Here is one way to provide a Paint(which only has a default constructor) object as a parameter:
  /// ```
  /// SpinnerWheel(
  ///       ..,
  ///       ..,
  ///       ..,
  ///       wheelPaint: Paint()
  ///             ..color = Colors.white
  ///             ..style = PaintingStyle.fill,
  ///       ..,
  ///),
  /// ```
  Spinwheel(
      {this.items,
      this.imageItems,
      @required this.select,
      this.size = 100.0,
      @required this.onChanged,
      this.rotationDuration = 200,
      this.touchToRotate = true,
      @required this.autoPlay,
      this.autoPlayDuration = const Duration(seconds: 5),
      this.longPressToPauseAutoplay = false,
      this.shouldDrawDividers = true,
      this.shouldDrawBorder = true,
      this.shouldDrawCenterPiece = true,
      this.shouldHighlight = true,
      this.hideOthers = true,
      this.highlightWhileRotating = false,
      this.wheelPaint,
      this.borderPaint,
      this.sectorDividerPaint,
      this.centerPiecePaint,
      this.highlightPaint,
      this.shutterPaint,
      this.wheelOrientation = 0,
      this.fadeDuration = const Duration(milliseconds: 1500),
      this.loadingIndicator})
      : itemCount = items != null ? items.length : imageItems.length,
        assert(
          (items != null && imageItems == null) ||
              (items == null && imageItems != null),
          'Spinwheel widget can only take either [items] or [imageItems] parameter'
          'and not both. Please remove one to clear this exception.',
        ),
        assert(
          (onChanged != null),
          'Spinwheel requires an onChange function. If it is not required,'
          'pass an empty function with a string parameter.',
        ),
        assert(select != null),
        assert(autoPlay != null);

  @override
  _SpinwheelState createState() => _SpinwheelState();
}

class _SpinwheelState extends State<Spinwheel>
    with SingleTickerProviderStateMixin {
  // The animation controller that controls the rotation of the spinwheel.
  AnimationController _rotationController;
  Animation _rotationAnimation;

  // To determine whether the pan on the wheel results in clockwise or
  // counter-clockwise motion.
  bool _movingClockwise = true;
  double _velocity;

  // Unified list for list of String and NamedImage.
  List<dynamic> _spinnerItems;

  // Used by components to check whether the given items are images or not.
  bool _isImageList = false;

  // List to store images loaded using ui.instantiateImageCodec()
  List<ui.Image> loadedImages = [];

  // Timer for autoplay.
  Timer _autoPlayTimer;

  Widget _loadingIndicator;

  @override
  void initState() {
    super.initState();
    initFields();
  }

  void initFields() {
    _rotationController = AnimationController(
        duration: Duration(milliseconds: widget.rotationDuration), vsync: this);

    _rotationAnimation =
        Tween(begin: 0.0, end: 1.0).animate(_rotationController)
          ..addListener(_rotationHandler)
          ..addStatusListener(_rotationStatusHandler);

    // Assigning to a unified list
    _spinnerItems = widget.items ?? widget.imageItems;
    _isImageList = widget.imageItems != null;

    // If image details are provided, load them.
    if (_isImageList) {
      Utilities.loadImages(widget.imageItems, context, loadedImages, () {
        notifySelectedOption();
      }, this);
    } else {
      Future.delayed(Duration(microseconds: 1), () {
        notifySelectedOption();
      });
    }

    _loadingIndicator = widget.loadingIndicator ??
        Center(
          child: FittedBox(
            child: Text(
              '...',
              style: TextStyle(fontSize: widget.size / 2),
            ),
          ),
        );

    autoPlayStart();
  }

  void _rotationHandler() {
    setState(() {});
  }

  void _rotationStatusHandler(AnimationStatus status) {
    if (status == AnimationStatus.completed) {
      if (_movingClockwise)
        previousItem();
      else
        nextItem();
      _rotationController.reset();
    }
  }

  // Function that triggers callback to set selected value.
  void notifySelectedOption() {
    if (_isImageList)
      widget.onChanged((_spinnerItems[widget.select] as NamedImage).name);
    else
      widget.onChanged(_spinnerItems[widget.select]);
  }

  @override
  Widget build(BuildContext context) {
    Widget wheel = FittedBox(
      fit: BoxFit.scaleDown,
      child: AnimatedOpacity(
        opacity: _isImageList
            ? (loadedImages.length == widget.imageItems.length) ? 1 : 0
            : 1,
        duration: widget.fadeDuration,
        child: GestureDetector(
          onTap: _tapHandler,
          onLongPress: _longPressHandler,
          onLongPressEnd: _longPressEndHandler,
          onPanUpdate: _panUpdateHandler,
          onPanEnd: _panEndHandler,
          child: CustomPaint(
            painter: SpinwheelPainter(
              widget.itemCount,
              _isImageList,
              _spinnerItems,
              _isImageList ? loadedImages : null,
              _movingClockwise,
              _rotationAnimation,
              widget.select,
              widget.shouldDrawDividers,
              widget.shouldDrawBorder,
              widget.shouldDrawCenterPiece,
              widget.hideOthers,
              widget.shouldHighlight,
              widget.highlightWhileRotating,
              widget.wheelPaint,
              widget.borderPaint,
              widget.sectorDividerPaint,
              widget.centerPiecePaint,
              widget.highlightPaint,
              widget.shutterPaint,
              widget.wheelOrientation,
            ),
            child: Container(
              width: widget.size,
              height: widget.size,
            ),
          ),
        ),
      ),
    );

    return _isImageList
        ? (loadedImages.length == widget.imageItems.length)
            ? wheel
            : _loadingIndicator
        : wheel;
  }

  // Tapping rotates the wheel to previously panned direction. Default is clockwise.
  void _tapHandler() {
    startAnimation();
    autoPlayRestart();
  }

  // Long press to pause
  void _longPressHandler() {
    if (widget.longPressToPauseAutoplay) autoPlayCancel();
  }

  // Start autoplay after long press ends
  void _longPressEndHandler(LongPressEndDetails details) {
    if (widget.longPressToPauseAutoplay) autoPlayStart();
  }

  // Function that keeps track of radial movement i.e, movement along the
  // wheel during each pan session and determines whether the pan
  // indicates clockwise or counter-clockwise motion.
  void _panUpdateHandler(DragUpdateDetails update) {
    bool onTop = update.localPosition.dy <= widget.size / 2;
    bool onRight = update.localPosition.dx >= widget.size / 2;
    bool onBottom = !onTop;
    bool onLeft = !onRight;

    bool panningUp = update.delta.dy <= 0;
    bool panningRight = update.delta.dx >= 0;
    bool panningDown = !panningUp;
    bool panningLeft = !panningRight;

    // Absolute change in the X and Y direction.
    double absX = update.delta.dx.abs();
    double absY = update.delta.dy.abs();

    bool isHorizontalClockwiseMotion =
        (onTop && panningRight) || (onBottom && panningLeft);
    bool isVerticalClockwiseMotion =
        (onLeft && panningUp) || (onRight && panningDown);

    double horizontalRotation = isHorizontalClockwiseMotion ? absX : -absX;
    double verticalRotation = isVerticalClockwiseMotion ? absY : -absY;

    // Resultant rotation. Can be used to find the velocity by multiplying
    // with update.delta.distance.
    double resultantRotation = verticalRotation + horizontalRotation;

    _movingClockwise = resultantRotation > 0;
    _velocity = resultantRotation * update.delta.distance;
  }

  // Function to be called at the end of each pan session and
  // trigger the rotation animation.
  void _panEndHandler(DragEndDetails details) {
    if (_rotationController.status != AnimationStatus.forward &&
        (widget.touchToRotate || _velocity != null)) {
      startAnimation();
      autoPlayRestart();
      _velocity = null;
    }

    autoPlayRestart();
  }

  // Cancel and start autoplay
  void autoPlayRestart() {
    autoPlayCancel();
    autoPlayStart();
  }

  void autoPlayStart() {
    if (widget.autoPlay) {
      _autoPlayTimer = Timer.periodic(widget.autoPlayDuration, (timer) {
        startAnimation();
      });
    }
  }

  // Function that starts the animation after a pan session.
  void startAnimation() async {
    await _rotationController.forward();
  }

  void autoPlayCancel() {
    if (_autoPlayTimer != null) if (widget.autoPlay &&
        _autoPlayTimer.isActive) {
      _autoPlayTimer.cancel();
    }
  }

  // Function that shifts the ordering of the menu list to the right
  // and triggers callback to set
  void previousItem() {
    Utilities.shiftItemsRight(_spinnerItems);
    if (_isImageList) Utilities.shiftItemsRight(loadedImages);
    notifySelectedOption();
  }

  // Function that shifts the ordering of the menu list to the left.
  void nextItem() {
    Utilities.shiftItemsLeft(_spinnerItems);
    if (_isImageList) Utilities.shiftItemsLeft(loadedImages);
    notifySelectedOption();
  }

  @override
  void dispose() {
    super.dispose();
    if (_rotationController != null) {
      _rotationController.dispose();
      _rotationController = null;
      _rotationAnimation.removeListener(_rotationHandler);
      _rotationAnimation.removeStatusListener(_rotationStatusHandler);
    }
    if (widget.autoPlay) _autoPlayTimer.cancel();
  }
}


lib/src/utilities.dart
import 'dart:typed_data';
import 'dart:ui' as ui;

import 'package:flutter/material.dart';

import 'named_image.dart';

class Utilities {
// Loading images from assetbundle using path from details in [imageItems].
  static void loadImages(List<NamedImage> imageItems, BuildContext context,
      List<ui.Image> loadedImages, Function notify, State state) async {
    ByteData byteData;
    Uint8List buffer;
    ui.Codec codec;
    ui.FrameInfo frameInfo;

    for (NamedImage namedImage in imageItems) {
      if (state.mounted) {
        byteData = await DefaultAssetBundle.of(context)
            .load(namedImage.path)
            .catchError((error) {
          throw FlutterError(
              'Error while loading image at ${namedImage.path}.. Check if the asset is'
              'specified in pubspec.yaml and that the path is correct. Error info: $error');
        });

        buffer = Uint8List.view(byteData.buffer);
        codec = await ui.instantiateImageCodec(buffer);

        frameInfo = await codec.getNextFrame();
        loadedImages.add(frameInfo.image);
      }
    }
    if (state.mounted) notify();
  }

  static void shiftItemsLeft(List<dynamic> items) {
    items.add(items.removeAt(0));
  }

  static void shiftItemsRight(List<dynamic> items) {
    items.insert(0, items.removeLast());
  }
}


lib/src/painter_presets.dart
import 'package:flutter/material.dart';

class PainterPresets {
  Paint get wheelPaintPreset {
    return Paint()
      ..style = PaintingStyle.fill
      ..color = Colors.grey.shade200;
  }

  Paint get borderPaintPreset {
    return Paint()
      ..style = PaintingStyle.stroke
      ..strokeWidth = 5.0
      ..color = Colors.grey.shade300;
  }

  Paint get sectorDividerPaintPreset {
    return Paint()
      ..style = PaintingStyle.stroke
      ..strokeWidth = 5.0
      ..color = Colors.white;
  }

  Paint get centerPiecePaintPreset {
    return Paint()
      ..style = PaintingStyle.fill
      ..color = Colors.red.shade300;
  }

  Paint get highlightPaintPreset {
    return Paint()
      ..style = PaintingStyle.fill
      ..color = Colors.black.withOpacity(0.2);
  }

  Paint get shutterPaintPreset {
    return Paint()
      ..style = PaintingStyle.fill
      ..color = Colors.grey.shade400;
  }
}


example/lib/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

import 'package:flutter_spinwheel/flutter_spinwheel.dart';
import 'package:carousel_slider/carousel_slider.dart';

List<Widget> getAppBarActions(BuildContext context) {
  return [
    FlatButton(
      child: Icon(
        Icons.device_unknown,
        color: Colors.white,
      ),
      onPressed: () {
        Navigator.pushReplacementNamed(context, '/sample');
      },
    ),
    FlatButton(
      child: Icon(
        Icons.insert_emoticon,
        color: Colors.white,
      ),
      onPressed: () {
        Navigator.pushReplacementNamed(context, '/emoji');
      },
    ),
    FlatButton(
      child: Icon(
        Icons.image,
        color: Colors.white,
      ),
      onPressed: () {
        Navigator.pushReplacementNamed(context, '/image');
      },
    ),
  ];
}

void main() => runApp(SpinwheelDemo());

class SpinwheelDemo extends StatefulWidget {
  @override
  _SpinwheelDemoState createState() => _SpinwheelDemoState();
}

class _SpinwheelDemoState extends State<SpinwheelDemo> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Spinwheel Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      initialRoute: '/emoji',
      routes: {
        '/sample': (context) => SpinwheelSampleApp(),
        '/emoji': (context) => SpinwheelEmojiDemo(),
        '/image': (context) => SpinwheelImageDemo()
      },
    );
  }
}

class SpinwheelSampleApp extends StatefulWidget {
  @override
  _SpinwheelSampleAppState createState() => _SpinwheelSampleAppState();
}

class _SpinwheelSampleAppState extends State<SpinwheelSampleApp> {
  List<String> questions;
  List<List<dynamic>> choices;
  List<String> answers;
  int select;
  String currentText;

  @override
  void initState() {
    super.initState();
    questions = [
      'Your first programming language?',
      'Did you own any pets?',
      'Choose any one vehicle you own from below',
    ];
    choices = [
      ['Dart', 'C', 'C++', 'Java', 'Python', 'JS', 'TS', '😠'],
      ['Yes', 'No'],
      [
        NamedImage(
          path: 'assets/images/car.jpg',
          name: 'car',
        ),
        NamedImage(
          path: 'assets/images/bike.jpg',
          name: 'bike',
        ),
        NamedImage(
          path: 'assets/images/other.jpg',
          name: 'other',
        ),
      ]
    ];
    select = 0;
    answers = ['', '', ''];
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
          title: Center(
            child: Text('Sample'),
          ),
          actions: getAppBarActions(context)),
      body: Scrollbar(
        child: ListView.builder(
            itemCount: 3,
            itemBuilder: (context, index) => Container(
                  margin: EdgeInsets.all(20.0),
                  height: MediaQuery.of(context).size.height / 2,
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.spaceAround,
                    children: <Widget>[
                      Text(questions[index]),
                      Text(answers[index] == '😠'
                          ? '${answers[index]} None of these'
                          : answers[index]),
                      Spinwheel(
                        items: choices[index][0] is String
                            ? choices[index].cast<String>()
                            : null,
                        imageItems: choices[index][0] is NamedImage
                            ? choices[index].cast<NamedImage>()
                            : null,
                        select: select,
                        autoPlay: false,
                        hideOthers: false,
                        shouldDrawBorder: false,
                        onChanged: (val) {
                          if (this.mounted)
                            setState(() {
                              answers[index] = val;
                            });
                        },
                      )
                    ],
                  ),
                )),
      ),
    );
  }
}

class SpinwheelEmojiDemo extends StatefulWidget {
  @override
  _SpinwheelEmojiDemoState createState() => _SpinwheelEmojiDemoState();
}

class _SpinwheelEmojiDemoState extends State<SpinwheelEmojiDemo> {
  List<String> items;
  int select;
  String currentText;

  @override
  void initState() {
    super.initState();
    items = ['🏘', '🚓', '🚛', '🏍', '🎉'];
    select = 2;
    currentText = items[select];
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
          title: Center(
            child: Text('Emoji'),
          ),
          actions: getAppBarActions(context)),
      body: Column(
        mainAxisAlignment: MainAxisAlignment.spaceAround,
        children: <Widget>[
          Center(
            child: Text(
              currentText,
              style: TextStyle(fontSize: 100.0),
            ),
          ),
          Spinwheel(
            items: items,
            onChanged: (val) {
              if (this.mounted)
                setState(() {
                  currentText = val;
                });
            },
            shouldHighlight: false,
            size: 150.0,
            select: 2,
            rotationDuration: 250,
            autoPlay: true,
            shouldDrawDividers: true,
          ),
        ],
      ),
    );
  }
}

class SpinwheelImageDemo extends StatefulWidget {
  @override
  _SpinwheelImageDemoState createState() => _SpinwheelImageDemoState();
}

class _SpinwheelImageDemoState extends State<SpinwheelImageDemo> {
  List<String> carouselNames;
  List<NamedImage> imgPack;
  CarouselSlider carousel;

  @override
  void initState() {
    super.initState();

    carouselNames = ['dog', 'cat', 'budgie', 'goldfish'];

    imgPack = [
      NamedImage(
        path: 'assets/images/dog.jpg',
        name: carouselNames[0],
        offsetX: 1.2,
      ),
      NamedImage(
        path: 'assets/images/cat.jpg',
        name: carouselNames[1],
        offsetY: 1.7,
      ),
      NamedImage(
        path: 'assets/images/budgie.jpg',
        name: carouselNames[2],
      ),
      NamedImage(
          path: 'assets/images/goldfish.jpg',
          name: carouselNames[3],
          offsetX: 1.2),
    ];

    // For Image example
    carousel = CarouselSlider(
      items: imgPack.map((img) {
        return Container(
          padding: EdgeInsets.all(10.0),
          color: Colors.orange,
          child: Image.asset(img.path),
        );
      }).toList(),
      autoPlay: false,
      enlargeCenterPage: true,
      viewportFraction: 1.0,
      aspectRatio: 2.0,
      initialPage: 2,
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Center(
          child: Text('Image'),
        ),
        actions: getAppBarActions(context),
      ),
      body: Column(
        mainAxisAlignment: MainAxisAlignment.spaceAround,
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: <Widget>[
          carousel,
          Center(
            child: Spinwheel(
              size: 250.0,
              imageItems: imgPack,
              select: 1,
              onChanged: (val) {
                carousel.animateToPage(carouselNames.indexOf(val),
                    duration: Duration(seconds: 1), curve: Curves.linear);
              },
              rotationDuration: 250,
              autoPlay: true,
              longPressToPauseAutoplay: true,
              hideOthers: false,
              shouldDrawDividers: true,
              shouldDrawBorder: false,
              shouldDrawCenterPiece: false,
              wheelOrientation: pi / 4,
            ),
          ),
        ],
      ),
    );
  }
}
